### DISCUSSION TITLE
Intermediate representation of Koka before compiling to C backend

### DISCUSSION THREAD

**Discussion by swarnpriya** [koka-lang/koka#478] [General]
Hi,

Is there an intermediate language (IR) of Koka before it is compiled to C? I would like to do some program analysis on the koka program and see things like constant-propagation / constant-folding etc. Or is it that all the optimisations are handled by the backend compiler?

For example: I want to ensure that division by 0 (static values) never occurs and the koka program is accepted only when we are sure that the divisor in non-zero. If there was an IR then I could write a pass like constant-propagation and do program analysis to find that divisor is never zero.

**Comment by TimWhiting** [comment]
Core is the IR

**Reply by TimWhiting** [reply]
BTW, I'm working on static analysis for Koka, I wonder if there is some overlap or aspects we could collaborate on.
